README.md
[![Build Status](https://travis-ci.org/SirTori/indenticator.svg?branch=master)](https://travis-ci.org/SirTori/indenticator)

# Indenticator

Visually highlights the current indent depth.

This extension can be used by itself, but it's recommended to use it alongside the built-in indent guides (setting **editor.renderIndentGuides**). These show thin grey lines on *every* indent, while this extension highlights the indent on the indent depth the cursor is currently on.

**Changes in Visual Studio Code 1.23.0**:<br/>
With Version 1.23.0 VS Code introduced a new built-in active indent guide. For an explanation of differences and tips regarding the new feature and this extension please see [remarks section](#remarks)

<table>
  <tr>
    <td><img src="img/demo.gif"></td>
  </tr>
  <tr>
    <td>Example using default settings</td>
  </tr>
</table>


## Feature Highlights

- **Peeking around the current indent block**: *Optionally* a hover can be added on the current indent marker to peek before and/or after the current indent block.<br/>
*To activate set **indenticator.showHover** to `true`*

  <table>
    <tr>
      <td><img src="img/r0.7.0/hover_with_highlight.png"></td>
      <td><img src="img/r0.7.0/hover_without_highlight.png"></td>
    </tr>
    <tr>
      <td>Example has <b>indenticator.hover.peekBack</b> & <b>indenticator.hover.peekForward</b> set to <code>1</code> and <b>indenticator.hover.highlight</b> set to <code>true</code></td>
      <td>Example has <b>indenticator.hover.peekBack</b> & <b>indenticator.hover.peekForward</b> set to <code>1</code> and <b>indenticator.hover.highlight</b> set to <code>false</code></td>
    </tr>
  </table>

- **Highlighting the the indent guide for the contained block**: *Optionally* a second indent guide can be configured to show the block *contained* by the current cursor position<br/>
*To activate set **indenticator.inner.showIndentGuide** to `true`*

  <table>
    <tr>
      <td><img src="img/r0.5.0/example_inner.png"></td>
    </tr>
    <tr>
      <td>Example has <b>indenticator.inner.showIndentGuide</b> set to <code>true</code> and <b>indenticator.inner.color.dark</b> set to <code>#ff000</code></td>
    </tr>
  </table>


- **Language specific settings**: The extension can be configured for each language separately to accommondate the requirements of different coding styles.

  Example:
    ``` JS
    {
      "indenticator.languageSpecific": {
        "[json, jsonc]": {
          //...
        },
        "[xml, html, xhtml]": {
          //...
        }
      }
    }
    ```

- **Highlight Styling**: All indent highlighting can be styled individually by defining color, width and border style.

## Settings and defaults
``` JS
{
    /******************************************************/
    /********** Standard Indent Highlight Config **********/
    /******************************************************/
    /* Whether to highlight the indent of the block enclosing the current line */
    "indenticator.showIndentGuide": true
    /* Color of the indent marker for dark themes */
    "indenticator.color.dark": "#888",
    /* Color of the indent marker for light themes */
    "indenticator.color.light": "#999",
    /* Width of the indent marker in pixels */
    "indenticator.width": 1,
    /* Line style of the indent marker (e.g. "inset", "dashed", "dotted", ...) */
    "indenticator.style": "inset",
    /* Whether to display the hover near the indent line */
    "indenticator.showHover": false,
    /* Wether to highlight the contained code block when hovering the indent line. If activated, the peeked content will be shown at top or bottom of the block, otherwise it will be shown at cursor position.*/
    "indenticator.hover.highlight": true,
    /* Lines before the current indent to be shown on hover */
    "indenticator.hover.peekBack": 1,
    /* Lines after the current indent to be shown on hover */
    "indenticator.hover.peekForward": 0,
    /* Remove lines from the hover at the beginning and end that have less characters than this */
    "indenticator.hover.trimLinesShorterThan": 2,
    /* Block placeholder to be written between peeked lines */
    "indenticator.hover.peekBlockPlaceholder": "...",

    /***************************************************/
    /********** Inner Indent Highlight Config **********/
    /***************************************************/
    /* Whether to highlight the indent of the block enclosed by the current line */
    "indenticator.inner.showHighlight": false
    /* Color of the indent marker for dark themes */
    "indenticator.inner.color.dark": "#888",
    /* Color of the indent marker for light themes */
    "indenticator.inner.color.light": "#999",
    /* Width of the indent marker in pixels */
    "indenticator.inner.width": 1,
    /* Line style of the indent marker (e.g. "inset", "dashed", "dotted", ...) */
    "indenticator.inner.style": "inset",
    /* Whether to display the hover near the inner indent line */
    "indenticator.inner.showHover": false,
    /* Wether to highlight the contained code block when hovering the indent line. If activated, the peeked content will be shown at top or bottom of the block, otherwise it will be shown at cursor position.*/
    "indenticator.inner.hover.highlight": true,
    /* Lines before the current inner indent to be shown on hover */
    "indenticator.inner.hover.peekBack": 1,
    /* Lines after the current inner indent to be shown on hover */
    "indenticator.inner.hover.peekForward": 0,
    /* Remove lines from the inner indent hover at the beginning and end that have less characters than this */
    "indenticator.inner.hover.trimLinesShorterThan": 2,
    /* Block placeholder to be written between peeked lines */
    "indenticator.inner.hover.peekBlockPlaceholder": "...",
    /* Whether to display the current indent depth on the statusbar */

    /********************************************/
    /********** Further Configurations **********/
    /********************************************/
    "indenticator.showCurrentDepthInStatusBar": true,
    /* A construct with language identifiers as properties containing a subset of indenticator options to be applied for that language */
    "indenticator.languageSpecific": {}
}
```
### Remarks
- **New Built-in active indent guide (VS Code 1.23.0)**
  - The built-in active indent guide always highlights the *highest* indent on the currently selected line, while this extensions highlights the indent depth the *cursor is currently at*. The difference can be seen in the following image:

    | ![Indenticator demonstration](img/r0.6.0/new_active_indent.png) |
    |- |
    | *yellow* is the new built-in active indent guide<br/>*red* is the inner active indent guide of this extension<br/>*grey* is the default active indent guide of this extension |

  - The built-in *active* indent guide can be configured through the setting: **editor.highlightActiveIndentGuide**

- **indenticator.languageSpecific**: The language identifiers can be viewed by using [VS Codes language selection](https://code.visualstudio.com/docs/languages/overview#_language-id) in the statusbar. To be analogous to [language specific editor options](https://code.visualstudio.com/docs/getstarted/settings#_language-specific-editor-settings) of VS Code the key has to be put between square brackets. But to simplify configuration of similar languages multiple keys can be listed within the square brackets as a comma seperated list.

  Any setting for indenticator that can be set normally, can also be set for the language specific configuration. If any setting is not set for the specific language the overall configuration will be used.

  Example:
  ``` JS
  {
    "indenticator.languageSpecific": {
      "[json, jsonc]": {
        "indenticator.showHover": true,
        "indenticator.hover.peekBack": 1
      }
    }
  }
  ```

CHANGELOG.md
# 0.7.0

## Minor
- Renamed settings:
  - **indenticator.showHighlight** changed to **indenticator.showIndentGuide**
  - **indenticator.inner.showHighlight** changed to **indenticator.inner.showIndentGuide**
- At some point the behaviour for hovers in VS Code changed slightly:

  Now if the hover highlights a code block the content of the hover is shown above or below the highlighted block, rather than at the cursor position. Without highlighting a code block the hover will still be shown at the cursor position.

  To give the user control over how the peek feature should behave when hovering on the indent guides new settings were added:
  - **indenticator.hover.highlight** (default `true`): Controls wether hovering on the indent guide should highlight the code block
  - **indenticator.inner.hover.highlight** (default `true`): Controls wether hovering on the inner indent guide should highlight the code block
- Hovering now works even if the indent guides themselves are disabled
## Patch
- Fixes for #23, #25 & #26
- Dependencies updated

# 0.6.0

## Minor

- Switched default border style from 'solid' to 'inset'
  - With VS Code 1.23.0 there seems to have been a change in the rendering of solid borders that makes them appear thicker than before. To offset that change the default style is now inset which should make the default borders somewhat slimmer again.
- Updated README to reflect recent changes with new buildin active indent guide of VS Code

## Patch
- additional checks before fetching lines from document to fix #20

# 0.5.1

## Path

- Fixed some typos

# 0.5.0

## Minor

- New Setting **indenticator.showHighlight** to toggle the Highlighting of the indent line. (default is `true`)
  - If this setting is `false` the hover options will be ignored
- New Feature: Highlighting for inner indent as suggested in #14
  - highlights the indent guide of the block contained by the current cursor position
  - added settings **indenticator.inner.\*** allowing for all configuration options already present for the standard indent highlight to be configured for the new highlight of the inner indent.
    - To keep the extensions behaviour for existing users **indenticator.inner.showHighlight** defaults to `false`

  ![Indenticator demonstration](img/r0.5.0/example_inner.png)

- Hover-Feature now also highlights the block it peeks around

  ![Indenticator demonstration](img/r0.5.0/example_highlight.png)

- Setting **indenticator.languageSpecific** now allows configuration of multiple languages at once by naming them in a comma seperated list

  Example:
    ``` JS
    {
      "indenticator.languageSpecific": {
        "[json, jsonc]": {
          //...
        },
        "[xml, html, xhtml]": {
          //...
        }
      }
    }
    ```

# 0.4.2

## Patch

- Fixed Issue #6: include first and last line of file in hover

# 0.4.1

## Patch

- Fixed Issue #5: Language specific settings will now be applied correctly

# 0.4.0

## Minor

- More Configuration Options for hover on indent line
  - changed setting **indenticator.showHover** to accept boolean and number. If given a number the hover is active if the current indent block is at list that many lines long. (default is `false`)
  - added setting **indenticator.hover.peekBack** to set the number of lines to be peeked before the current indent block (default is `1`)
  - added setting **indenticator.hover.peekForward** to set the number of lines to be peeked after the current indent block (default is `0`)
  - added setting **indenticator.hover.trimLinesShorterThan** to remove lines at beginning and end of hover if they are shorter then the given value (default is `2`)
  - added setting **indenticator.hover.peekBlockPlaceholder** to set string to be shown between the peeked lines in the hover (default is "`...`")
- Added option to specify language specific configurations
  - added setting **indenticator.languageSpecific** to specify language specific settings (default is `{}`)

    Accepts an object with language keys enclosed in square brackets (analogous to [language specific editor options](https://code.visualstudio.com/docs/getstarted/settings#_language-specific-editor-settings)) as property keys and any set of indenticator configuartion as values.

    Example:
    ``` JS
    {
      "indenticator.languageSpecific": {
        "[json]": {
          "indenticator.showHover": true,
          "indenticator.hover.peekBack": 1
        }
      }
    }
    ```

# 0.3.0

## Minor
- New Feature: Hover on active indent line peeks first line (thanks to [rsbondi](https://github.com/rsbondi) on github)
  - added setting **indenticator.showHover** to activate the new Feature (default is `false`)

## Patch
- Updated project to Typescript 2
- Fixed Travis builds

# 0.2.1

## Patch

- Fixed Issue #1: The Extension will no longer add multiple Information Items into the Statusbar

# 0.2.0

## Minor

- First release to be published to the VS Marketplace

test/stubs.ts
import * as vscode from 'vscode';

export class TextLine {
    public firstNonWhitespaceCharacterIndex: number;
    public isEmptyOrWhitespace: boolean;
    public lineNumber: number;
    public range: vscode.Range;
    public rangeIncludingLineBreak: vscode.Range;
    public text: string;
}

export class TextDocument {
    public fileName: string;
    public isDirty: boolean;
    public isUntitled: boolean;
    public lineCount: number;
    public uri: vscode.Uri;
    public version: number;

    public argReturnStub = (arg?: any) => {return arg;};

    public getTextStub = this.argReturnStub;
    public getText(range?: vscode.Range) : string {
        return this.getTextStub(range);
    }

    public getWordRangeAtPositionStub = this.argReturnStub;
    public getWordRangeAtPosition(position: vscode.Position) : vscode.Range {
        return this.getWordRangeAtPositionStub(position);
    }

    public lineAtStub = this.argReturnStub;
    public lineAt(line: number) : vscode.TextLine;
    public lineAt(line: vscode.TextLine) : vscode.TextLine;
    public lineAt(line: any) : vscode.TextLine {
        return this.lineAtStub(line);
    }

    public offsetAtStub = this.argReturnStub;
    public offsetAt(position: vscode.Position) : number {
        return this.offsetAtStub(position);
    }

    public positionAtStub = this.argReturnStub;
    public positionAt(offset: number) : vscode.Position{
        return this.positionAtStub(offset);
    }

    public saveStub = this.argReturnStub;
    public save() : Thenable<boolean> {
        return this.saveStub();
    }

    public validatePositionStub = this.argReturnStub;
    public validatePosition(position: vscode.Position) {
        return this.validatePositionStub(position);
    }

    public validateRangeStub = this.argReturnStub;
    public validateRange(range: vscode.Range) : vscode.Range {
        return this.validateRangeStub(range);
    }
}


test/runTest.ts
import * as path from "path";
import { runTests } from 'vscode-test';

async function go() {
    try {

        const extensionDevelopmentPath = path.resolve(__dirname, '../../')
        const extensionTestsPath = path.resolve(__dirname, './suite')
        const testWorkspace = path.resolve(__dirname, '../../test-fixtures')
        runTests({
            extensionDevelopmentPath,
            extensionTestsPath: extensionTestsPath,
            launchArgs: [testWorkspace]
        });
    } catch (err) {
       console.error('Failed to run tests');
       process.exit(1);
    }
}

go();

src/extension.ts
'use strict';
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import {window, Disposable, ExtensionContext, StatusBarAlignment,
        StatusBarItem, TextDocument, TextEditor, TextEditorOptions,
        TextEditorDecorationType, TextLine, Selection, Range,
        Position, workspace, env, languages, WorkspaceConfiguration, Hover
} from 'vscode';
// this method is called when your extension is activated
// your extension is activated the very first time the command is executed
export function activate(context: ExtensionContext) {

    let indentSpy = new IndentSpy();
    let indentSpyController = new IndentSpyController(indentSpy);

    context.subscriptions.push(indentSpy);
    context.subscriptions.push(indentSpyController);
}

// this method is called when your extension is deactivated
export function deactivate() {
}

class LanguageConfig {
    constructor(public langConfig: any,
                public config: WorkspaceConfiguration) {}

    get<T>(name: string, defaultValue?:T ): T {
        let v = this.langConfig[`indenticator.${name}`];
        if(v !== undefined) {
            return v;
        }
        return this.config.get(name, defaultValue);
    }
}

class IndentConfiguration {
    show: boolean;
    style: TextEditorDecorationType;
    hover: number;
    hoverConf: {
        highlight: boolean,
        peekBack: number,
        peekForward: number,
        trimLinesShorterThan: number,
        peekBlockPlaceholder: string
    };
    hoverProvider: Disposable
    firstLine: number;
    lastLine: number;
    indentPos: number;
}

export class IndentSpy {
    _locales: Object;
    _currentLocale: Object;
    _statusBarItem: StatusBarItem;
    _outerConf: IndentConfiguration = new IndentConfiguration();
    _innerConf: IndentConfiguration = new IndentConfiguration();

    constructor() {
        this._locales = {
            en: {statusText: `Indents: {indent}`,
                 statusTooltip: `current indent depth: {indent}`},
            de: {statusText: `Einzüge: {indent}`,
                 statusTooltip: `aktuelle Einzugtiefe: {indent}`},
            ja: {statusText: `字下げ: {indent}`,
                 statusTooltip: `現在のインデントの深さ: {indent}`},
            default: {statusText: `Indents: {indent}`,
                      statusTooltip: `current indent depth: {indent}`},
        };
        this.updateConfig();
    }

    public updateConfig() {
        this._clearDecorators();

        let locale = env.language;
        let multipartLocale = env.language.indexOf('-');
        if(multipartLocale >= 0) {
            locale = locale.substring(0, multipartLocale);
        }

        if(!this._locales[locale]) {
            this._currentLocale = this._locales['default'];
        } else {
            this._currentLocale = this._locales[locale];
        }
        let langConfig = {};
        let config = workspace.getConfiguration('indenticator');
        if(window.activeTextEditor) {
            let docLang = window.activeTextEditor.document.languageId
            let allLangConfig = config.get("languageSpecific", {});
            let docLangKey = Object.keys(allLangConfig).find(k => {
                return k.match(`^\\[(.*,\\s*)?${docLang}(,.*)?\\]$`) !== null;
            });
            if(docLangKey) {
                langConfig = allLangConfig[docLangKey] || {};
            }
        }
        let myConf = new LanguageConfig(langConfig, config);

        if(myConf.get('showCurrentDepthInStatusBar')) {
            if(!this._statusBarItem) {
                this._statusBarItem = window.createStatusBarItem(
                    StatusBarAlignment.Right, 100);
            }
        } else if(this._statusBarItem) {
            this._statusBarItem.dispose();
            this._statusBarItem = undefined;
        }

        this._outerConf.show = myConf.get('showIndentGuide');
        this._innerConf.show = myConf.get('inner.showIndentGuide');

        this._outerConf.style = window.createTextEditorDecorationType({
            dark: {
                borderColor: myConf.get('color.dark', '#888'),
                borderStyle: myConf.get('style', 'inset'),
                borderWidth: myConf.get('width', 1) + "px"
            },
            light: {
                borderColor: myConf.get('color.light', '#999'),
                borderStyle: myConf.get('style', 'inset'),
                borderWidth: myConf.get('width', 1) + "px"
            }
        });

        this._innerConf.style = window.createTextEditorDecorationType({
            dark: {
                borderColor: myConf.get('inner.color.dark', '#888'),
                borderStyle: myConf.get('inner.style', 'inset'),
                borderWidth: myConf.get('inner.width', 1) + "px"
            },
            light: {
                borderColor: myConf.get('inner.color.light', '#999'),
                borderStyle: myConf.get('inner.style', 'inset'),
                borderWidth: myConf.get('inner.width', 1) + "px"
            }
        });

        let showHover:boolean|number = myConf.get('showHover', false);
        if(typeof showHover === 'boolean') {
            this._outerConf.hover = showHover ? 1 : 0;
        } else {
            this._outerConf.hover = showHover;
        }
        if(this._outerConf.hover) {
            this._outerConf.hoverConf = {
                highlight: myConf.get('hover.highlight', true),
                peekBack: myConf.get('hover.peekBack', 1),
                peekForward: myConf.get('hover.peekForward', 0),
                trimLinesShorterThan: myConf.get(
                    'hover.trimLinesShorterThan', 2),
                peekBlockPlaceholder: myConf.get(
                    'hover.peekBlockPlaceholder', '...')
            };
        } else if (this._outerConf.hoverProvider) {
            this._outerConf.hoverProvider.dispose();
        }

        showHover = myConf.get('inner.showHover', false);
        if(typeof showHover === 'boolean') {
            this._innerConf.hover = showHover ? 1 : 0;
        } else {
            this._innerConf.hover = showHover;
        }
        if(this._innerConf.hover) {
            this._innerConf.hoverConf = {
                highlight: myConf.get('inner.hover.highlight', true),
                peekBack: myConf.get('inner.hover.peekBack', 1),
                peekForward: myConf.get('inner.hover.peekForward', 0),
                trimLinesShorterThan: myConf.get(
                    'inner.hover.trimLinesShorterThan', 2),
                peekBlockPlaceholder: myConf.get(
                    'inner.hover.peekBlockPlaceholder', '...')
            };
        } else if (this._innerConf.hoverProvider) {
            this._innerConf.hoverProvider.dispose();
        }

        this.updateCurrentIndent();
    }

    public updateCurrentIndent() {
        let hideStatusbarIfPossible = () => {
            if(this._statusBarItem) {
                this._statusBarItem.hide();
            }
        }

        let editor = window.activeTextEditor;
        if (!editor) {
            hideStatusbarIfPossible();
            return;
        }

        let document = editor.document;
        if (!document) {
            hideStatusbarIfPossible();
            return;
        }

        let selection = editor.selection;
        if (!selection) {
            hideStatusbarIfPossible();
            return;
        }

        let tabSize = this._getTabSize(editor.options);
        let selectedIndent = this._getSelectedIndentDepth(document, selection, tabSize);
        if(this._outerConf.show || this._outerConf.hover || this._innerConf.show || this._innerConf.hover) {
            let activeRanges = this._getActiveIndentRanges(document, selection, selectedIndent, tabSize);
            if(this._outerConf.show) {
                editor.setDecorations(this._outerConf.style, activeRanges.outer);
            }
            if(this._outerConf.hover && activeRanges.outer.length >= this._outerConf.hover) {
                this._buildHover(editor, tabSize, this._outerConf);
            } else if(this._outerConf.hoverProvider) {
                this._outerConf.hoverProvider.dispose();
            }
            if(this._innerConf.show) {
                editor.setDecorations(this._innerConf.style, activeRanges.inner);
            }
            if(this._innerConf.hover && activeRanges.inner.length >= this._innerConf.hover) {
                this._buildHover(editor, tabSize, this._innerConf);
            } else if(this._innerConf.hoverProvider) {
                this._innerConf.hoverProvider.dispose();
            }
        }

        if(this._statusBarItem){
            this._statusBarItem.text = this._currentLocale['statusText']
                .replace('{indent}', selectedIndent);
            this._statusBarItem.tooltip = this._currentLocale['statusTooltip']
                .replace('{indent}', selectedIndent);
            this._statusBarItem.show();
        }
    }

    _buildHover(editor: TextEditor, tabSize: number, conf: IndentConfiguration) {
        if (conf.hoverProvider) {
            conf.hoverProvider.dispose();
        }
        conf.hoverProvider = languages.registerHoverProvider(
            editor.document.languageId,
            {
                provideHover: (doc, position) => {
                    return this._buildHoverprovider(position, editor, tabSize, conf);
                }
            }
        );
    }

    _buildHoverprovider(position: Position, editor: TextEditor,
                        tabSize: number, conf: IndentConfiguration): Hover {
        let char = conf.indentPos
        if(position.character > char - 1
           && position.character < char + 1
           && position.line >= conf.firstLine
           && position.line <= conf.lastLine) {
            let str = this._buildHoverString(editor, tabSize, conf);
            let range;
            if (conf.hoverConf.highlight) {
                range = new Range(conf.firstLine, conf.indentPos,
                                  conf.lastLine, conf.indentPos)
            }
            if(str) {
                return {
                    range: range,
                    contents: [
                        {
                            language: editor.document.languageId,
                            value: str
                        }
                    ]
                };
            }
            return null;
        }
    }

    _buildHoverString(editor: TextEditor, tabSize: number,
                      conf: IndentConfiguration): string {
        let hoverLines = [];
        let document = editor.document;
        let refDepth = this._getLinesIndentDepth(
            document.lineAt(conf.firstLine), tabSize);

        let backHoverLines = this._peekBack(editor.document, tabSize, refDepth, conf);
        let forwardHoverLines = this._peekForward(editor.document, tabSize, refDepth, conf);

        hoverLines.push(...backHoverLines);
        if(forwardHoverLines.length > 0 || backHoverLines.length > 0) {
            hoverLines.push(this._buildHoverPlaceholder(editor, tabSize, conf));
        }
        hoverLines.push(...forwardHoverLines);
        return hoverLines.join('\n');
    }

    _buildHoverPlaceholder(editor: TextEditor, tabSize: number,
                           conf: IndentConfiguration): string {
        let tabChar = editor.options.insertSpaces?' ':'\t';
        let spacing = tabChar.repeat(tabSize);
        return `${spacing}${conf.hoverConf.peekBlockPlaceholder}`;
    }

    _peekBack(document: TextDocument, tabSize: number,
              refDepth: number, conf: IndentConfiguration): Array<string> {
        let backHoverLines = [];
        if(conf && conf.hoverConf && conf.hoverConf.peekBack > 0) {
            let firstPeekLine = Math.max(
                conf.firstLine - (conf.hoverConf.peekBack - 1), 0);
            let pushedOnce = false;
            for(let i = firstPeekLine; i <= conf.firstLine; i++) {
                let line = document.lineAt(i)
                let lineStr = line.text.trim();
                if(!pushedOnce &&
                   lineStr.length < conf.hoverConf.trimLinesShorterThan) {
                    continue;
                }
                let lineDepth = this._getLinesIndentDepth(line, tabSize);
                if(lineDepth != refDepth) {
                    backHoverLines.splice(0);
                    continue;
                }
                backHoverLines.push(lineStr);
                pushedOnce = true;
            }
        }
        return backHoverLines;
    }

    _peekForward(document: TextDocument, tabSize: number,
                 refDepth: number, conf: IndentConfiguration): Array<string> {
        let forwardHoverLines = [];
        if(conf && conf.hoverConf && conf.hoverConf.peekForward > 0) {
            let lastPeekLine = Math.min(
                conf.lastLine + (conf.hoverConf.peekForward - 1),
                document.lineCount - 1);
            let pushedOnce = false;
            for(let i = lastPeekLine; i >= conf.lastLine; i--) {
                let line = document.lineAt(i)
                let lineStr = line.text.trim();
                if(!pushedOnce &&
                   lineStr.length < conf.hoverConf.trimLinesShorterThan) {
                    continue;
                }
                let lineDepth = this._getLinesIndentDepth(line, tabSize);
                if(lineDepth != refDepth) {
                    forwardHoverLines.splice(0);
                    continue;
                }
                forwardHoverLines.push(lineStr);
                pushedOnce = true;
            }
        }
        return forwardHoverLines.reverse();
    }

    _clearDecorators() {
        if(this._outerConf.style) {
            for(let i = 0; i < window.visibleTextEditors.length; i++) {
                window.visibleTextEditors[i].setDecorations(
                    this._outerConf.style, []);
            }

        }
        if(this._innerConf.style) {
        for(let i = 0; i < window.visibleTextEditors.length; i++) {
                window.visibleTextEditors[i].setDecorations(
                    this._innerConf.style, []);
            }

        }
    }

    _getTabSize(options: TextEditorOptions) {
        return options.insertSpaces?Number(options.tabSize):1;
    }

    _getIndentDepth(index: number, tabSize: number) {
        return Math.ceil(index / tabSize);
    }

    _getLinesIndentDepth(line: TextLine, tabSize: number) {
        return this._getIndentDepth(line.firstNonWhitespaceCharacterIndex,
                                    tabSize);
    }

    _createIndicatorRange(line: number, character: number) {
        return new Range(new Position(line, character),
                         new Position(line, character));
    }

    _getSelectedIndentDepth(document: TextDocument, selection: Selection,
                            tabSize: number) {
        if(selection.isSingleLine) {
            let maxlineNum = document.lineCount - 1;
            let line = document.lineAt(Math.min(selection.start.line, maxlineNum));
            return this._getIndentDepth(
                Math.min(selection.start.character,
                         line.firstNonWhitespaceCharacterIndex),
                tabSize);
        }
        let selectedIndent = Number.MAX_VALUE;
        let maxlineNum = Math.min(selection.end.line,document.lineCount - 1);
        for(let i = selection.start.line; i <= maxlineNum; i++) {
            let line = document.lineAt(i);
            if(line.isEmptyOrWhitespace) {
                continue;
            }
            selectedIndent = Math.min(selectedIndent,
                                      this._getLinesIndentDepth(line, tabSize));
        }
        return selectedIndent;
    }

    _getActiveIndentRanges(document: TextDocument, selection: Selection,
                           selectedIndent: number, tabSize: number){
        let activeRanges = [];
        let activeInnerRanges = [];
        let line: TextLine;
        let innerDeactivated: boolean;

        this._outerConf.firstLine = selection.start.line;
        this._outerConf.lastLine = selection.end.line;
        this._outerConf.indentPos = (selectedIndent - 1) * tabSize;

        this._innerConf.firstLine = selection.start.line;
        this._innerConf.lastLine = selection.end.line;
        this._innerConf.indentPos = selectedIndent * tabSize;

        let addRanges = (i: number, line: TextLine) => {
            let lineAdded = false;
            let innerAdded = false;
            let lineIndent = this._getLinesIndentDepth(line, tabSize);
            if(!innerDeactivated && (
                    lineIndent > selectedIndent || (
                        line.isEmptyOrWhitespace && selectedIndent === lineIndent &&
                        (i !== selection.end.line || selection.end.character !== this._innerConf.indentPos)))) {
                activeInnerRanges.push(
                    this._createIndicatorRange(i, this._innerConf.indentPos));
                lineAdded = true;
                innerAdded = true;
        }
            if(this._outerConf.indentPos >= 0 && (
                    lineIndent >= selectedIndent || (
                        line.isEmptyOrWhitespace && selectedIndent === 1))) {
                activeRanges.push(this._createIndicatorRange(i, this._outerConf.indentPos));
                lineAdded = true;
        }
            return {
                'lineAdded': lineAdded,
                'innerAdded': innerAdded
            };
        };

        // add ranges for preceeding lines on same indent
        innerDeactivated = false;
        for(let i = selection.start.line; i >= 0; i--) {
            line = document.lineAt(i);
            let result = addRanges(i, line)
            if(!result.innerAdded && !line.isEmptyOrWhitespace && !innerDeactivated) {
                innerDeactivated = true;
                this._innerConf.firstLine = i;
            }
            if(!result.lineAdded && !line.isEmptyOrWhitespace) {
                this._outerConf.firstLine = i;
                break;
            }
        }
        // add ranges for following lines on same indent
        innerDeactivated = false;
        for(let i = selection.start.line + 1; i < document.lineCount; i++) {
            line = document.lineAt(i);
            let result = addRanges(i, line)
            if(!result.innerAdded && !line.isEmptyOrWhitespace && !innerDeactivated) {
                innerDeactivated = true;
                this._innerConf.lastLine = i;
            }
            if(!result.lineAdded && !line.isEmptyOrWhitespace) {
                this._outerConf.lastLine = i;
                break;
            }
        }
        return {
            outer: activeRanges,
            inner: activeInnerRanges
        };
    }

    dispose() {
        if(this._statusBarItem){
            this._statusBarItem.dispose();
        }
        if(this._outerConf.hoverProvider) {
            this._outerConf.hoverProvider.dispose();
        }
        if (this._innerConf.hoverProvider) {
            this._innerConf.hoverProvider.dispose();
        }
    }
}

class IndentSpyController {

    private _indentSpy: IndentSpy;
    private _disposable: Disposable;

    constructor(indentSpy: IndentSpy) {
        this._indentSpy = indentSpy;
        this._indentSpy.updateCurrentIndent();

        // subscribe to selection change and editor activation events
        let subscriptions: Disposable[] = [];
        window.onDidChangeTextEditorSelection(
            this._onUpdateEvent, this, subscriptions);
        window.onDidChangeActiveTextEditor(
            this._onChangedEditor, this, subscriptions);

        // subscribe to configuration change events
        workspace.onDidChangeConfiguration(
            this._onChangedConfigEvent, this, subscriptions);

        this._disposable = Disposable.from(...subscriptions);
    }

    dispose() {
        this._disposable.dispose();
    }

    private _onUpdateEvent(e) {
        this._indentSpy.updateCurrentIndent();
    }

    private _onChangedEditor(e) {
        this._indentSpy.updateConfig();
        this._indentSpy.updateCurrentIndent();
    }

    private _onChangedConfigEvent(e) {
        this._indentSpy.updateConfig();
    }
}